agent/processwatcher.go

```diff
@@ -0,0 +1,434 @@
+package agent
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"os"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/appcanary/libspector"
+)
+
+type ProcessWatcher interface {
+	Start()
+	Stop()
+	Match() string
+	StateJson() []byte
+}
+
+// Types
+
+// Singleton control object
+type processWatcher struct {
+	sync.Mutex
+	keepPolling  bool
+	UpdatedAt    time.Time
+	OnChange     ChangeHandler
+	pollSleep    time.Duration
+	BeingWatched bool
+	match        string
+	stateJson    []byte
+	checksum     uint32
+}
+
+// process objects with references to systemLibraries
+type systemProcesses []watchedProcess
+
+// indexed list of libraries
+type systemLibraries []libspector.Library
+
+// This is how the data structure should look
+
+// var someLib1 libspector.Library
+// var someLib2 libspector.Library
+// var processStartTime time.Time
+// var updatedAt time.Time
+// var callback = func(w Watcher) { return }
+
+// var watcher = processWatcher{
+// 	keepPolling: true,
+// 	UpdatedAt:   updatedAt,
+// 	OnChange:    callback,
+// 	pollSleep:   env.PollSleep,
+// 	match:       "",
+//  // this is no longer true, instead we just keep the marshaled json, but this
+//  // a helpful structure doc
+// 	state: &processMap{
+// 		processes: &systemProcesses{
+// 			watchedProcess{
+// 				ProcessStartedAt: processStartTime,
+// 				Libraries: []library{
+// 					processLibrary{
+// 						Outdated:     true,
+// 						libraryIndex: 2, // somelib2
+// 					},
+// 				},
+// 				Outdated: true,
+// 				Pid:      5,
+// 				Command:  "test_script.sh 1 2 3",
+// 			},
+// 			// ...
+// 		},
+// 		libraries: &systemLibraries{
+// 			someLib1,
+// 			someLib2,
+// 			// ...
+// 		},
+// 	},
+// }
+
+// "map" in the colloquial sense
+type processMap struct {
```

Comment:
what if instead of 'processMap' we called this 'systemState'? I also spent a bunch of time parsing out the diff b/w processMap and processWatcher.